src/core/maps/area_occupancy_estimator.h

```diff
@@ -81,9 +81,68 @@ class AreaOccupancyEstimator : public CellOccupancyEstimator {
   virtual Occupancy estimate_occupancy(const Beam &beam,
                                        const Rectangle &cell_bnds,
                                        bool is_occ) override {
-    Intersections intrs = find_intersections(beam, cell_bnds, is_occ);
-    double chunk_area = compute_chunk_area(beam, cell_bnds, is_occ, intrs);
-    return estimate_occupancy(chunk_area, cell_bnds.area(), is_occ);
+    Occupancy result(0.0, 0.0);
+    if (equal (beam.x_st, beam.x_end) && (equal(beam.y_st, beam.y_end))) {
+      result.setInvalid();
+      return result;
+    }
+    if(!beam.intersects_rect(cell_bnds)) {
+      result.setInvalid();
+      return result;
+    }
+    if(beam.encounters_rect(cell_bnds)){
+      return is_occ ? Occupancy(1.0, 1.0) : Occupancy(base_empty_prob(), 0.5);
+    }
+    if(!cell_bnds.contains(beam.x_end, beam.y_end) && is_occ) {
+      result.setInvalid();
+      return result;
+    }
+    Beam buf_beam = beam.generate_revert();
```

Comment:
But it is definitely not a buffer. Why can't you just write

beam.revert().encounters_rect(...)